src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs

```diff
@@ -0,0 +1,176 @@
+				return;
+
+			if (!FieldHasInvalidSerializeField(symbols, out var fieldName))
+				return;
+
+			context.ReportDiagnostic(Diagnostic.Create(Rule, field.GetLocation(), fieldName));
+		}
+
+		private static bool PropertyHasInvalidSerializeField(ISymbol symbol, out string propertyName)
+		{
+			propertyName = null;
+			if (!(symbol is IPropertySymbol))
+				return false;
+
+			var containingType = symbol.ContainingType;
+			if (!containingType.Extends(typeof(UnityEngine.Object)))
+				return false;
+
+			bool hasSerializeFieldAttribute = symbol
+				.GetAttributes()
+				.Any(a => a.AttributeClass.Matches(typeof(UnityEngine.SerializeField)));
+
+			if (!hasSerializeFieldAttribute)
+				return false;
+
+			propertyName = symbol.Name;
+			return true;
+		}
+
+		private static bool FieldHasInvalidSerializeField(List<ISymbol> symbols, out string fieldName)
+		{
+			fieldName = null;
+
+			bool foundMatchingSymbol = false;
+			foreach (var symbol in symbols)
+			{
+				if (!(symbol is IFieldSymbol && symbol.DeclaredAccessibility == Accessibility.Public))
+					continue;
+
+				var containingType = symbol.ContainingType;
+				if (!containingType.Extends(typeof(UnityEngine.Object)))
+					continue;
+
+				if (!symbol.GetAttributes().Any(a => a.AttributeClass.Matches(typeof(UnityEngine.SerializeField))))
+					continue;
+
+				fieldName = symbol.Name;
+				foundMatchingSymbol = true;
+				break;
+			}
+
+			return foundMatchingSymbol;
+		}
+	}
+
+	[ExportCodeFixProvider(LanguageNames.CSharp)]
+	public class ImproperSerializeFieldCodeFix : CodeFixProvider
+	{
+		public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(ImproperSerializeFieldAnalyzer.Id);
+
+		public sealed override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+		public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+		{
+			var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+			var declaration = root.FindNode(context.Span) as MemberDeclarationSyntax;
+			if (declaration == null)
+				return;
+
+			if (!(declaration is PropertyDeclarationSyntax || declaration is FieldDeclarationSyntax))
+				return;
+
+			context.RegisterCodeFix(
+				CodeAction.Create(
+					Strings.ImproperSerializeFieldCodeFixTitle,
+					ct => RemoveSerializeFieldAttribute(context.Document, declaration, ct),
+					declaration.ToFullString()),
+				context.Diagnostics);
+		}
+
+		private static async Task<Document> RemoveSerializeFieldAttribute(Document document, MemberDeclarationSyntax declaration, CancellationToken cancellationToken)
+		{
+			var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+			var attributes = new SyntaxList<AttributeListSyntax>();
+
+			foreach (var attributeList in declaration.AttributeLists)
+			{
+				var nodes = attributeList
+					.Attributes
+					.Where(a => a.Name.ToString() == "SerializeField");
```

Comment:
Let's avoid comparing names. We want to make sure that we're removing the proper attribute so we need to work at the symbol level and enforce that it's actually a UnityEngine.SerializeField, like you do in the diagnostic above.